porch/engine/pkg/engine/k8sruntime.go

```diff
@@ -0,0 +1,170 @@
+		kubeClient: kubeClient,
+		namespace:  namespace,
+	}, nil
+}
+
+type kubeFunctionRuntime struct {
+	// kubeClient is the kubernetes client
+	kubeClient client.Client
+
+	// namespace holds the namespace where the executors run
+	namespace string
+}
+
+var _ kpt.FunctionRuntime = &kubeFunctionRuntime{}
+
+func (k *kubeFunctionRuntime) GetRunner(ctx context.Context, fn *v1.Function) (fn.FunctionRunner, error) {
+	image := fn.Image
+
+	tokens := strings.SplitN(image, ":", 2)
+	if len(tokens) != 2 {
+		// TODO: Assume latest?
+		return nil, fmt.Errorf("expected version in image %q", image)
+	}
+
+	functionName := lastComponent(tokens[0])
+
+	// Using labels here is a WIP experiment.
+	// In theory it allows a pod/image to support multiple functions ... but that raises a discovery/trust issues
+	// It might be easier simply to look at the image that the pod is running
+	// However, today this is handy because it lets us run functions from a mirror, i.e. we can
+	// give them a stable name in the package, but push them to our per-project gcr.io mirror.
+	// TODO: Review use of labels if/when we start auto-launching pods.
+
+	matchLabels := map[string]string{
+		"functions.porch.kpt.dev/" + functionName: "",
+	}
+	// Work around limits of annotation/label keys/values
+	matchAnnotations := map[string]string{
+		"functions.porch.kpt.dev/" + functionName: image,
+	}
+
+	var pods corev1.PodList
+
+	var options []client.ListOption
+	options = append(options, client.InNamespace(k.namespace))
+	options = append(options, client.MatchingLabels(matchLabels))
+	if err := k.kubeClient.List(ctx, &pods, options...); err != nil {
+		return nil, fmt.Errorf("error listing pods: %w", err)
+	}
+
+	// TODO: we should launch/manage the pods, rather than trusting the labels
+	for i := range pods.Items {
+		pod := &pods.Items[i]
+		if pod.Status.Phase != "Running" {
+			continue
+		}
+
+		matchesAnnotations := true
+		for k, v := range matchAnnotations {
+			if pod.Annotations[k] != v {
+				matchesAnnotations = false
+				break
+			}
+		}
+		if !matchesAnnotations {
+			break
+		}
+		return &kubeFunctionRunner{
+			pod:   pod,
+			image: image,
+		}, nil
+	}
+
+	return nil, fmt.Errorf("could not find pod to run function %q", image)
+}
+
+func lastComponent(s string) string {
+	lastSlash := strings.LastIndex(s, "/")
+	return s[lastSlash+1:]
+}
+
+func (k *kubeFunctionRuntime) Close() error {
+	return nil
+}
+
+type kubeFunctionRunner struct {
+	pod   *corev1.Pod
+	image string
+}
+
+var _ fn.FunctionRunner = &kubeFunctionRunner{}
+
+func (k *kubeFunctionRunner) Run(r io.Reader, w io.Writer) error {
+	// We shouldn't be putting this into the runner
+	ctx := context.TODO()
+
+	address := k.pod.Status.PodIP
+	if address == "" {
+		return fmt.Errorf("pod did not have podIP")
+	}
+	address += ":8888"
+
+	klog.Infof("dialing grpc function runner %q", address)
+
+	// TODO: pool connections
+	cc, err := grpc.Dial(address, grpc.WithTransportCredentials(insecure.NewCredentials()))
```

Comment:
is the intention for this to work for arbitrary functions?
I understand we can use image rewriting (which will have issues with image signatures later on), including an example of a concrete function would help.

Reply:
Yes, this works for arbitrary functions - if they implement the GRPC protocol.  We _could_ take a non-enlightened function and just overlay your existing function-runner onto it.  But technically it's possible that a function might not be safe to run in parallel / multiple times in the same pod.  I don't know if we want to require that functions are safe to run in parallel / multiple times (maybe supporting an opt-out mechanism).  If we're going to do that, maybe we should just require them to build in GRPC support!

Reply:
let's track this as a follow-up.

Reply:
I'm working a PoC that has a different flavor of pod evaluator. It doesn't force every KRM function to support gRPC.
It 1) creates pod 2) evaluates function 3) brings it down. More concretely, it does the following:
- Create a pod with 1 regular container (containing the real KRM function) and 1 initContainer (containing a gRPC wrapper server). These 2 containers share the emptyDir volume.
- In initContainer, it copies a wrapper gRPC server binary to the emptyDir volume.
- In the regular container, we run the gPRC server which will invoke the KRM function binary.

I should be able to create a draft PR today.